journal/src/main/java/io/zeebe/journal/file/record/SBESerializer.java

```diff
@@ -0,0 +1,137 @@
+
+  private final MessageHeaderDecoder headerDecoder = new MessageHeaderDecoder();
+  private final RecordMetadataDecoder metadataDecoder = new RecordMetadataDecoder();
+  private final RecordDataDecoder recordDecoder = new RecordDataDecoder();
+
+  @Override
+  public int writeData(
+      final RecordData record, final MutableDirectBuffer buffer, final int offset) {
+    if (offset + getSerializedLength(record) > buffer.capacity()) {
+      throw new BufferOverflowException();
+    }
+
+    headerEncoder
+        .wrap(buffer, offset)
+        .blockLength(recordEncoder.sbeBlockLength())
+        .templateId(recordEncoder.sbeTemplateId())
+        .schemaId(recordEncoder.sbeSchemaId())
+        .version(recordEncoder.sbeSchemaVersion());
+
+    recordEncoder.wrap(buffer, offset + headerEncoder.encodedLength());
+
+    recordEncoder
+        .index(record.index())
+        .asqn(record.asqn())
+        .putData(record.data(), 0, record.data().capacity());
+
+    return headerEncoder.encodedLength() + recordEncoder.encodedLength();
+  }
+
+  @Override
+  public int writeMetadata(
+      final RecordMetadata metadata, final MutableDirectBuffer buffer, final int offset) {
+
+    headerEncoder
+        .wrap(buffer, offset)
+        .blockLength(metadataEncoder.sbeBlockLength())
+        .templateId(metadataEncoder.sbeTemplateId())
+        .schemaId(metadataEncoder.sbeSchemaId())
+        .version(metadataEncoder.sbeSchemaVersion());
+
+    metadataEncoder.wrap(buffer, offset + headerEncoder.encodedLength());
+
+    metadataEncoder.checksum(metadata.checksum()).length(metadata.length());
+
+    return headerEncoder.encodedLength() + metadataEncoder.encodedLength();
+  }
+
+  @Override
+  public int getMetadataLength() {
+    return headerEncoder.encodedLength() + metadataEncoder.sbeBlockLength();
+  }
+
+  @Override
+  public boolean hasMetadata(final DirectBuffer buffer, final int offset) {
+    headerDecoder.wrap(buffer, offset);
+    return (headerDecoder.schemaId() == metadataDecoder.sbeSchemaId()
+        && headerDecoder.templateId() == metadataDecoder.sbeTemplateId());
+  }
+
+  @Override
+  public RecordMetadata readMetadata(final DirectBuffer buffer, final int offset) {
+    if (!hasMetadata(buffer, offset)) {
+      throw new InvalidRecordException("Cannot read buffer. Header does not match.");
+    }
+    metadataDecoder.wrap(
+        buffer,
+        offset + headerDecoder.encodedLength(),
+        headerDecoder.blockLength(),
+        headerDecoder.version());
+
+    return new RecordMetadata(metadataDecoder.checksum(), metadataDecoder.length());
+  }
+
+  @Override
+  public RecordData readData(final DirectBuffer buffer, final int offset, final int length) {
+    headerDecoder.wrap(buffer, offset);
+    if (!(headerDecoder.schemaId() == recordDecoder.sbeSchemaId()
+        && headerDecoder.templateId() == recordDecoder.sbeTemplateId())) {
+      throw new InvalidRecordException("Cannot read buffer. Header does not match.");
```

Comment:
Same here. I think this isn't being handled

Reply:
If there is a valid metadata, then it is expected to have a valid record at this position. If not then the log is most likely corrupted. How should we handle this exception?
Would you suggest to ignore the exception in the reader, and treat it as end of the log?

Reply:
Actually, checksum is compared before call to this method. If the data is corrupted, it would be detected by the checksum. Then we are already treating it as end of the log. So ideally this exception should never happen, even if the log is corrupted.